### ISSUE TITLE
Generate generic variable names in a less naive way

### ISSUE THREAD

**Issue by brendanzab** [Marwes/gluon#74]
See https://github.com/Marwes/gluon/issues/58#issuecomment-234927599 and #72

```haskell
{
    ...
    foldl: (t0 -> t1 -> t0) -> t0 -> std.prelude.List t1 -> t0,
    foldr: (t2 -> t3 -> t3) -> t3 -> std.prelude.List t2 -> t3,
    make_Monoid: std.prelude.Monoid t4 -> {
        append: t5 -> t5 -> t5,
        empty: t6,
        <>: t7 -> t7 -> t7
    },
    ...
    <<: (t100 -> t101) -> (t102 -> t100) -> t102 -> t101,
    >>: (t103 -> t104) -> (t104 -> t105) -> t103 -> t105,
    ...
}
```

**Comment by Marwes** [comment]
Another thing I have forgotten to open an issue about :disappointed: 

Relevant code https://github.com/Marwes/gluon/blob/de751ebece1ab252af07320ac1aafd21b0070c17/check/src/typecheck.rs#L1054-L1061

**Comment by bjz** [comment]
Heh, no worries! That's why us contributors are here! ^_^

**Comment by Marwes** [comment]
I have seen generated generic variables such as `a0000000001` so the prelude example isn't even that bad :smile: . The current generator is just super naive.

**Comment by Marwes** [comment]
And I see now that `make_Monoid` expresses the exact bug that RankNTypes are needed for. Need to extract that as a test case!

**Comment by Marwes** [comment]
Changed the title as the pretty printing is only a symptom of the problem

**Comment by bjz** [comment]
Yeah, I had no idea how to express it reasonably. That is much better!

**Comment by Marwes** [comment]
Started working on a fix for this as the same tracking that is needed to fix this is needed for #77
